pkg/labels: Fix missing source field in policy labels

pillai-ashwin opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer's Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

## Description

This PR fixes an issue where CiliumNetworkPolicy labels with empty `source` fields were not being assigned a default source value, causing issues in policy debugging and Hubble flow exports.

### Problem
When parsing CiliumNetworkPolicy labels from YAML like:
```yaml
labels:
- key: policy-comment
  value: allow all traffic inside namespace
```

Labels with empty source fields were passed through without any default source assignment, resulting in:
- `cilium-dbg policy get` showing empty source fields: `"source": ""`
- Hubble flow exports showing labels starting with `:` instead of `unspec:` (e.g., `:policy-comment=allow all traffic inside namespace`)
- Policy correlation and debugging difficulties

### Root Cause
The `ParseToCiliumLabels` function in `pkg/k8s/apis/cilium.io/utils/utils.go` was not handling labels with empty source fields. Since CNP `Parse` can be called directly from the policy directory watcher (not just from Kubernetes resources), we cannot assume the label source is always `k8s`.

### Solution
- Modified `ParseToCiliumLabels()` to convert empty source fields to `LabelSourceUnspec` ("unspec")
- Labels with `unspec` source remain as `unspec` (unchanged)
- Labels with explicit sources (k8s, custom, etc.) are preserved as-is
- Added comprehensive unit tests to verify the correct source assignment behavior

### Main idea
Labels with empty source will be set to 'unspec' and will be internally handled as `LabelSourceAny` when required for matching. This ensures consistent behavior across both Kubernetes-sourced policies and directory watcher-sourced policies.

### Testing
- All existing tests pass
- New test case "parse labels with empty source" verifies correct behavior
- Labels now correctly appear as `"unspec:policy-comment=allow all traffic inside namespace"`

Fixes: #40912

```release-note
Fix CiliumNetworkPolicy labels without explicit source field defaulting to empty string instead of "unspec", resolving issues in policy debugging and Hubble flow exports.
```

christarazi reviewed (commented):
I think the fix makes sense, but stepping back; shouldn't the source be `k8s` instead of `unspec` if the label being added was done so through k8s labels? I'm not saying the default should be `k8s`, but why isn't the `policy-comment` label that is being picked up by Cilium detected as coming from `k8s` to begin with?

fristonio commented on pkg/labels/labels.go:
```diff
@@ -474,7 +474,12 @@ func (l *Label) UnmarshalJSON(data []byte) error {
 
-		l.Source = aux.Source
+		// If source is empty, default to LabelSourceUnspec
+		if aux.Source == "" {
```
Thanks for the fix.
While this seems like a reasonable change, for the issue described this is not what we expect the behavior to be. If the policy is coming from k8s we should assign the source label as K8s instead of unspec.

We parse CiliumNetworkPolicy from k8s to cilium API rule representation here:
https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/utils/utils.go#L383
https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/v2/cnp_types.go#L163
We should make sure that the label source is set to k8s here.

pillai-ashwin replied:
Thanks! done.

squeed replied:
Are we sure this is right? We previously had a bug where Cilium network policy objects were not JSON round-trip-able. In other words, if you parse a JSON / YAML to the `CiliumNetworkPolicy` type, then re-export to YAML, the result was distinct (and invalid for submission to the apiserver). I would prefer JSON unmarshal-marshal to have no effect.

fristonio replied:
Thats a good point. However, in this particular case the Label itself is represented as a [struct of source,key,value fields](https://github.com/cilium/cilium/blob/main/pkg/labels/labels.go#L170) so it is round-trip-able(this is cilium internal label not k8s). But it does mean that Unmarshal/Marshal on the Label will yield a different result(Source as Unspec vs empty) but functionally its idempotent as we treat both the same way.
I don't have a strong opinion, but given we have had issues in the past with custom marshalling logic, I would also lean towards not changing the semantics during unmarshal.

@pillai-ashwin sorry for the back and forth on this, but what are your thoughts?

fristonio replied:
As discussed in community meeting @pillai-ashwin I think we want to avoid this custom Unmarshal behavior for cilium api types. Lets remove the changes here and assume that correct source will be populated once the CNP is parsed.

pillai-ashwin replied:
Thanks for the review @fristonio! Made changes -
* Removed the custom unmarshaling behavior that defaulted empty sources to "unspec"
* Removed test cases that tested the UnmarshalJSON defaulting behavior

fristonio commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -399,5 +399,17 @@ func ParseToCiliumLabels(namespace, name string, uid types.UID, ruleLbs labels.L
+
+	// Ensure user-defined labels from CiliumNetworkPolicy have k8s source
+	userLbls := make(labels.LabelArray, len(ruleLbs))
+	for i, lbl := range ruleLbs {
```
Thanks for the fix. I think we should still keep the old fix you had earlier for `UnmarshalJSON` along with this change. https://github.com/cilium/cilium/commit/948be07daf2e20c863fb1e5b4b5ebd173cd60087
When parsing cilium K8s rule we can check for both Upsec and empty source to populate source as K8s here.

pillai-ashwin replied:
Done!

fristonio commented on pkg/k8s/apis/cilium.io/utils/policy_label_test.go:
```diff
@@ -0,0 +1,63 @@
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+// TestPolicyLabelSourceFix tests the fix for issue #40912
```
nit: I don't think we need to reference the issue here.

fristonio commented on pkg/k8s/apis/cilium.io/utils/policy_label_test.go:
```diff
@@ -0,0 +1,63 @@
+// TestPolicyLabelSourceFix tests the fix for issue #40912
+// where CiliumNetworkPolicy labels without explicit source field
+// were getting empty source instead of being set to "k8s" source.
+func TestPolicyLabelSourceFix(t *testing.T) {
```
nit:
```suggestion
func TestParseToCiliumLabels(t *testing.T) {
```

pillai-ashwin replied:
There's already a `TestParseToCiliumLabels` function in `utils_test.go` so can't do that.

fristonio replied:
I think it makes more sense to extend the existing test cases in `TestParseToCiliumLabels` instead of creating a new test(or file). This test is already testing the codepath but needs these additional scenario to cover the changes in this PR.

pillai-ashwin replied:
Yes, that would be easier too. Moved the test cases to `TestParseToCiliumLabels`.

fristonio reviewed (changes requested):
One comment and a minor nit.

fristonio reviewed (approved):
Thanks for the contribution. LGTM <3

squeed reviewed (approved):
Nice!

pillai-ashwin committed:
pkg/k8s: Fix missing source field for CiliumNetworkPolicy labels

When CiliumNetworkPolicy labels have an empty source field, they should be
assigned 'unspec' source. This is because CNP Parse can be called directly
from policy directory watcher, not just from Kubernetes resources, so it's
incorrect to assume the label source is always k8s.

Labels with empty source will be set to 'unspec' and will be internally
handled as LabelSourceAny when required for matching.

This fixes the issue where policy labels appear with missing source in
Hubble flow exports and cilium-dbg policy get output.

Changes:
- Modified ParseToCiliumLabels to convert only empty sources to unspec
- Labels with unspec source remain unspec
- Labels with explicit sources (k8s, custom, etc.) are preserved as-is
- Updated test case to verify the correct source assignment behavior

Signed-off-by: Ashwin Pillai <pillai.ashwin@gmail.com>

fristonio commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -395,5 +395,17 @@ func ParseToCiliumLabels(namespace, name string, uid types.UID, ruleLbs labels.L
+	// Ensure user-defined labels from CiliumNetworkPolicy have k8s source
+	userLbls := make(labels.LabelArray, len(ruleLbs))
+	for i, lbl := range ruleLbs {
+		if lbl.Source == "" || lbl.Source == labels.LabelSourceUnspec {
```
Hey @pillai-ashwin 
Looking at the changes again, I missed that CNP `Parse` can be called directly from policy directory watcher as well. In that case, its incorrect to assume that the label source is k8s by default if not specified. I think the correct behavior will be to set the label source as unspec if empty. This will be internally handled as `LabelSourceAny` when required for matching.
Sorry for the back and forth on this.

pillai-ashwin replied:
Thanks for the review! Hmm yeah - CNP `Parse` can be called from the policy directory watcher, so we can't assume k8s source.

```go
if lbl.Source == "" {
    // If source is empty, set it to unspec
    userLbls[i] = labels.NewLabel(lbl.Key, lbl.Value, labels.LabelSourceUnspec)
} else {
    userLbls[i] = lbl
}
```

This will now ensure correct behavior for both Kubernetes-sourced CNPs and directory watcher-sourced CNPs. The labels with empty source will be internally handled as `LabelSourceAny` when needed for matching.
